Sources/SecureXPC/Server/XPCClientRequirement.swift

```diff
@@ -8,88 +8,364 @@
+        
+        // Validate the each component in the client path is present in the parent bundle path in the same order
+        for i in 0..<parentBundleURL.pathComponents.count {
+            if parentBundleURL.pathComponents[i] != clientPath.pathComponents[i] {
+                return false
+            }
+        }
+        
+        return true
+    }
+    
+    private func SecCodeCopyPath(_ code: SecCode) -> URL? {
+        var staticCode: SecStaticCode?
+        guard SecCodeCopyStaticCode(code, SecCSFlags(), &staticCode) == errSecSuccess,
+              let staticCode = staticCode else {
+            return nil
+        }
+        
+        var path: CFURL?
+        guard Security.SecCodeCopyPath(staticCode, SecCSFlags(), &path) == errSecSuccess,
+              let path = (path as URL?)?.standardized else {
+            return nil
+        }
+        
+        return path
+    }
+    
+    func isEqual(to acceptor: MessageAcceptor) -> Bool {
+        guard let acceptor = acceptor as? ParentBundleMessageAcceptor else {
+            return false
+        }
+        
+        return acceptor.parentBundleURL == self.parentBundleURL
+    }
+}
+
+/// Logically ands the results of two message acceptors.
+fileprivate struct AndMessageAcceptor: MessageAcceptor {
+    private let lhs: MessageAcceptor
+    private let rhs: MessageAcceptor
+    
+    init(lhs: MessageAcceptor, rhs: MessageAcceptor) {
+        self.lhs = lhs
+        self.rhs = rhs
+    }
+    
+    func acceptMessage(connection: xpc_connection_t, message: xpc_object_t) -> Bool {
+        lhs.acceptMessage(connection: connection, message: message) &&
+        rhs.acceptMessage(connection: connection, message: message)
+    }
+    
+    func isEqual(to acceptor: MessageAcceptor) -> Bool {
+        guard let acceptor = acceptor as? AndMessageAcceptor else {
+            return false
+        }
+        
+        return self.lhs.isEqual(to: acceptor.lhs) && self.rhs.isEqual(to: acceptor.rhs)
+    }
+}
+
+/// Logically ors the results of two message acceptors.
+fileprivate struct OrMessageAcceptor: MessageAcceptor {
+    private let lhs: MessageAcceptor
+    private let rhs: MessageAcceptor
+    
+    init(lhs: MessageAcceptor, rhs: MessageAcceptor) {
+        self.lhs = lhs
+        self.rhs = rhs
+    }
+    
+    func acceptMessage(connection: xpc_connection_t, message: xpc_object_t) -> Bool {
+        lhs.acceptMessage(connection: connection, message: message) ||
+        rhs.acceptMessage(connection: connection, message: message)
+    }
+    
+    func isEqual(to acceptor: MessageAcceptor) -> Bool {
+        guard let acceptor = acceptor as? OrMessageAcceptor else {
+            return false
+        }
+        
+        return self.lhs.isEqual(to: acceptor.lhs) && self.rhs.isEqual(to: acceptor.rhs)
+    }
+}
+
+/// If sandboxed, expands the sandbox by using a URL contained within the client's request.
+private func expandSandboxIfNecessary(message: xpc_object_t) throws {
```

Comment:
I'm so happy that you know all of this niche information, because I sure don't~ (until now!)

Reply:
Well I only known it now through reading lots of stuff, asking some questions, and some trial and error. XPC is rather arcane.